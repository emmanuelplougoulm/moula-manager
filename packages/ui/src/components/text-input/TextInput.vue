<template>
  <div class="container">
    <LabelItem
      :text="label"
      :highlighted="isHighlighted"
      :tooltip-title="props.tooltipTitle"
      :tooltip-description="props.tooltipDescription"
    />
    <input
      id="text-input"
      v-model="inputValue"
      data-testid="text-input"
      type="text"
      :placeholder="placeholder"
      :class="inputClass"
      class="text-big-regular input--default"
    />
    <CaptionItem :text="caption" />
    <ErrorItem :text="errorText" />
  </div>
</template>

<script setup lang="ts">
import { ref, watch } from 'vue';
import ErrorItem from '../error/ErrorItem.vue';
import CaptionItem from '../caption/CaptionItem.vue';
import LabelItem from '../label/LabelItem.vue';
import { getErrorText } from '../../utils/commons';

const props = defineProps({
  label: {
    type: String,
    required: true
  },
  placeholder: {
    type: String,
    default: ''
  },
  modelValue: {
    type: Object,
    required: true
  },
  pattern: {
    type: RegExp,
    default: /./
  },
  required: {
    type: Boolean,
    required: true
  },
  tooltipTitle: {
    type: String,
    default: undefined
  },
  tooltipDescription: {
    type: String,
    default: undefined
  },
  caption: {
    type: String,
    default: undefined
  }
});

let errorText = '';
const emit = defineEmits(['update:modelValue']);
const inputValue = ref(props.modelValue.value);
const isValid = ref(props.pattern.test(inputValue.value));
const isHighlighted = ref(props.required && !isValid.value);
let inputClass = props.required && !isValid.value ? 'input--highlighted' : '';

watch(inputValue, (newValue: string) => {
  isValid.value = props.pattern.test(newValue);

  if (props.required) {
    isHighlighted.value = !isValid.value;

    if (isHighlighted.value) {
      errorText = getErrorText(props.required, newValue);
      inputClass = 'input--highlighted';
    } else {
      errorText = '';
      inputClass = '';
    }
  }
  if (!props.required) {
    if (!isValid.value) {
      errorText = getErrorText(props.required, newValue);
    } else {
      errorText = '';
    }
  }
  emit('update:modelValue', { value: newValue, isValid: isValid.value });
});
</script>

<style scoped>
.container {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

input:focus-visible {
  outline: 0;
  background-color: var(--neutral-70);
  border: none;
  overflow: hidden;
  box-shadow: none;
}

.input--default {
  height: 3rem;
  padding: 0.75rem 1rem;
  border-radius: 0.25rem;
  border: none;
  background-color: var(--neutral-80);
  box-sizing: border-box;
}

.input--highlighted {
  border: 0.125rem var(--primary-30) solid;
  background-color: var(--primary-90);
  color: var(--primary-30);
}
</style>
