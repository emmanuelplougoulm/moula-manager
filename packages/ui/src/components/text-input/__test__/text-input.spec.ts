import { mount } from '@vue/test-utils';
import TextInput from '../TextInput.vue';

const textInput = '#text-input';

describe('TextInput tests', () => {
  it('should display highlighted class on label and text-area', () => {
    const wrapper = mount(TextInput, {
      props: {
        label: 'Description',
        required: true,
        pattern: /^([ \w]{1,128})(?![^?!§&#{[@\]}=+])/,
        modelValue: { value: '', isValid: undefined }
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display highlighted class on label and text-area with a caption', () => {
    const wrapper = mount(TextInput, {
      props: {
        label: 'Description',
        required: true,
        pattern: /^([ \w]{1,128})(?![^?!§&#{[@\]}=+])/,
        modelValue: { value: '', isValid: undefined },
        caption: 'This is a caption'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display default class on label and focus pseudo class on input', async () => {
    const wrapper = mount(TextInput, {
      props: {
        label: 'Informations',
        required: true,
        pattern: /^([ \w]{1,128})(?![^?!§&#{[@\]}=+])/,
        modelValue: { value: '2, place Citron', isValid: true }
      }
    });

    const input = wrapper.find(textInput);
    await input.setValue('2, place Citron');

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display highlighted class on label due to invalid format', async () => {
    const wrapper = mount(TextInput, {
      props: {
        label: 'Adresse complète',
        required: true,
        pattern: /^([ \w]{1,128})(?![^?!§&#{[@\]}=+])/,
        modelValue: { value: '', isValid: undefined }
      }
    });

    const input = wrapper.find(textInput);
    await input.setValue('*ù');

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display default class on label and text-area', async () => {
    const wrapper = mount(TextInput, {
      props: {
        label: 'Nom de famille',
        required: false,
        pattern: /^([ \w]{1,128})(?![^?!§&#{[@\]}=+])/,
        modelValue: { value: '', isValid: undefined }
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should not display highlighted class and error due to invalid format with the default pattern', async () => {
    const wrapper = mount(TextInput, {
      props: {
        label: 'Veuillez préciser',
        required: true,
        modelValue: { value: '', isValid: undefined }
      }
    });

    const input = wrapper.find(textInput);
    await input.setValue('*ù');

    expect(wrapper.html()).toMatchSnapshot();
  });
});
