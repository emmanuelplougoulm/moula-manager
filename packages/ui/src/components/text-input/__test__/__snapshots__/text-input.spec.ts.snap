// Vitest Snapshot v1

exports[`TextInput tests > should display default class on label and focus pseudo class on input 1`] = `
"<div data-v-e10f066f=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-e10f066f=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Informations <!--v-if--></span><input data-v-e10f066f=\\"\\" id=\\"text-input\\" data-testid=\\"text-input\\" type=\\"text\\" placeholder=\\"\\" class=\\"input--highlighted text-big-regular input--default\\">
  <!--v-if-->
  <!--v-if-->
</div>"
`;

exports[`TextInput tests > should display default class on label and text-area 1`] = `
"<div data-v-e10f066f=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-e10f066f=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default\\" data-testid=\\"label-item\\">Nom de famille <!--v-if--></span><input data-v-e10f066f=\\"\\" id=\\"text-input\\" data-testid=\\"text-input\\" type=\\"text\\" placeholder=\\"\\" class=\\"text-big-regular input--default\\">
  <!--v-if-->
  <!--v-if-->
</div>"
`;

exports[`TextInput tests > should display highlighted class on label and text-area 1`] = `
"<div data-v-e10f066f=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-e10f066f=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Description <!--v-if--></span><input data-v-e10f066f=\\"\\" id=\\"text-input\\" data-testid=\\"text-input\\" type=\\"text\\" placeholder=\\"\\" class=\\"input--highlighted text-big-regular input--default\\">
  <!--v-if-->
  <!--v-if-->
</div>"
`;

exports[`TextInput tests > should display highlighted class on label and text-area with a caption 1`] = `
"<div data-v-e10f066f=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-e10f066f=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Description <!--v-if--></span><input data-v-e10f066f=\\"\\" id=\\"text-input\\" data-testid=\\"text-input\\" type=\\"text\\" placeholder=\\"\\" class=\\"input--highlighted text-big-regular input--default\\"><span data-v-b41f83c4=\\"\\" data-v-e10f066f=\\"\\" data-testid=\\"caption-item\\" class=\\"text-small-regular\\">This is a caption</span>
  <!--v-if-->
</div>"
`;

exports[`TextInput tests > should display highlighted class on label due to invalid format 1`] = `
"<div data-v-e10f066f=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-e10f066f=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Adresse complète <!--v-if--></span><input data-v-e10f066f=\\"\\" id=\\"text-input\\" data-testid=\\"text-input\\" type=\\"text\\" placeholder=\\"\\" class=\\"input--highlighted text-big-regular input--default\\">
  <!--v-if--><span data-v-c1d8433f=\\"\\" data-v-e10f066f=\\"\\" id=\\"error-item\\" data-testid=\\"error-item\\" class=\\"text-small-regular\\">Invalid format</span>
</div>"
`;

exports[`TextInput tests > should not display highlighted class and error due to invalid format with the default pattern 1`] = `
"<div data-v-e10f066f=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-e10f066f=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default\\" data-testid=\\"label-item\\">Veuillez préciser <!--v-if--></span><input data-v-e10f066f=\\"\\" id=\\"text-input\\" data-testid=\\"text-input\\" type=\\"text\\" placeholder=\\"\\" class=\\"text-big-regular input--default\\">
  <!--v-if-->
  <!--v-if-->
</div>"
`;
